### ISSUE TITLE
hasura + timescaledb on pg11 > Upserts not working anymore

### ISSUE THREAD

**Issue by waterdrop01** [hasura/graphql-engine#1711]
Hi there,

I recently switched Hasura backend db from `timescale/timescaledb:1.2.1-pg10` to `timescale/timescaledb:1.2.1-pg11`.

As described [in this issue on Timescaledb github](https://github.com/timescale/timescaledb/issues/1094), when doing an **UPSERT** operation on an hypertable (through Graphql-engine), I'm now faced with this error: `ERROR:  hypertables do not support ON CONFLICT statements that reference constraints`.

To workaround this, Guys at Timescaledb advised me to instead of mentioning the constraint name (like does graphql-engine statement) :
```sql
INSERT INTO "data" ( "timestamp", "uuid" ) VALUES ( "2019-03-06T16:46:46.988Z", "9db6dbe1-cd14-1d17-38c0-4d4a36679fd6" ) ON CONFLICT ON CONSTRAINT "data_pkey" DO NOTHING RETURNING *
```
To explicitly mention column names (involved in the constraint):
![screenshot from 2019-03-07 08-48-46](https://user-images.githubusercontent.com/47316824/53940382-dcd95680-40b5-11e9-92ad-e44fbb574ccf.png)

Even though I believe the "design issue" is more on Timescaledb side, I'd be interested to hear your input on this > **would that make sense for Hasura to change the way it's working now, If not, why?**

Many thanks!

**Comment by 0x777** [comment]
> would that make sense for Hasura to change the way it's working now, If not, why?

This wouldn't be a right way to solve this problem. What if some other extension only supports using constraint names? Hasura should ideally switch between flavors of postgres (and their versions) and work around their limitations. However this looks slightly more complicated as we also need to check if a table is a hypertable, and then switch based on that.

**Comment by waterdrop01** [comment]
Hello @0x777 , thank you for answer. I just got feedback from the Timescale team on this issue:

> It is often preferable to use unique index inference rather than naming a constraint directly using ON CONFLICT ON CONSTRAINT constraint_name. Inference will continue to work correctly when the underlying index is replaced by another more or less equivalent index in an overlapping way, for example when using CREATE UNIQUE INDEX ... CONCURRENTLY before dropping the index being replaced.

and also:

> My suggestion would be that you use ... ON CONFLICT ("uuid", "timestamp")... .

This issue is actually not related to PG version, but instead to the latest TimescaleDB version (it comes as a pg extension).

**Since 2 months, Hasura GQL engine doesn't work with timescaledb anymore** (for UPSERT operations). It's too bad as these two projects are both gaining lot of traction. Any take from the Hasura team on how to solve this issue? 

FYI, here is the [dedicated thread on TimescaleDB github.](https://github.com/timescale/timescaledb/issues/1094#issuecomment-488372604)

Thank you!

**Comment by 0x777** [comment]
@waterdrop01 We'll try to add a workaround.

**Comment by waterdrop01** [comment]
@0x777  wow that would be great! really looking forward a solution to this, thanks.

**Comment by waterdrop01** [comment]
Hi @0x777 , congrats on the 1.0 beta0.1 release.

 I'd love to have timescale support in CA :)

I noticed this project has the exact same issue: https://github.com/JasperFx/marten/issues/1262

As mentioned: "According to the Postgres docs: It is often preferable to use unique index inference rather than naming a constraint directly using ON CONFLICT ON CONSTRAINT constraint_name."

Thanks, have a nice day

**Comment by waterdrop01** [comment]
Hello @0x777 , any chance to have that on the roadmap?

**Comment by waterdrop01** [comment]
@marionschleifer thanks for your help on this. @dsandip please ask any question, I'd be happy to assist you on this issue!

**Comment by waterdrop01** [comment]
Hi Hasura team, would that make sense that one of you @0x777 @shahidhk @dsandip @marionschleifer go on the [Timescale github issue](https://github.com/timescale/timescaledb/issues/1094), and share with them your opinion on how to fix this?

That would potentially bring compatibility back between Timescale and Hasura! (it's been 5 months it's broken).

**Comment by waterdrop01** [comment]
Hi @marionschleifer , any update on this issue? Thanks!

**Comment by waterdrop01** [comment]
`timescale/timescaledb:1.7.1-pg12` and `asura/graphql-engine:v1.2.1` are still not working correctly together :(
```
2020-05-26 20:39:36.663 UTC [1589] ERROR:  hypertables do not support ON CONFLICT statements that reference constraints
2020-05-26 20:39:36.663 UTC [1589] HINT:  Use column names to infer indexes instead.
```

**Comment by waterdrop01** [comment]
Hi @tirumaraiselvan, any visibility regarding this issue? Would be greatly appreciated

**Comment by haf** [comment]
https://github.com/hasura/graphql-engine/issues/1711#issuecomment-489552148 @0x777 What was the workaround?

**Comment by sulliwane** [comment]
Very much interested in the workaround as well! Thanks @0x777 @tirumaraiselvan for your input

**Comment by ondrejspilkaABB** [comment]
Hello, any resolution to this issue?

**Comment by waterdrop01** [comment]
Hello, @0x777 @marionschleifer @tirumaraiselvan, did you guys plan to release a fix to this bug? It's soon going to be 2 years this issue is open. Reproducing should be quite straightforward:

1. docker pull timescale/timescaledb:latest-pg12
2. docker pull hasura/graphql-engine
3. Connect to timescaledb (which is a postgres instance + an extension) and run `CREATE EXTENSION IF NOT EXISTS timescaledb CASCADE;` as well as `CREATE SCHEMA "schema1";` & `CREATE TABLE "schema1"."table1" ("id" int AS IDENTITY PRIMARY KEY,"value" text NOT NULL,  "timestamp" timestamp with time zone NOT NULL);` and finally run `SELECT create_hypertable('schema1.table1', 'timestamp');` to convert schema.table1 to a timescale "hypertable".
4. Send this mutation through Hasura:
```gql
mutation upsert_schema1_table1(
    $objects: [schema1_table1_insert_input!]!
    $update_columns: [schema1_table1_update_column!]!
  ) {
    insert_schema1_table1(
      objects: $objects
      on_conflict: { constraint: table1_pkey, update_columns: $update_columns }
    ) {
      returning {
        id
        value
       timestamp
      }
    }
  }
```
with these query variables:
```json
{
  "objects": {
    "id": 1,
    "value": "value",
    "timestamp": "2021-02-26T12:00:00+00:00",
  },
  "update_columns": ["id", "value", "timestamp"]
}
```
This will throw an "ON CONFLICT ON CONSTRAINT" error, as with Hypertables one can't reference a primary key by its name, but only by [its column names](https://docs.timescale.com/latest/using-timescaledb/writing-data#upsert):

![Screenshot from 2021-02-27 11-01-02](https://user-images.githubusercontent.com/47316824/109384182-5b56e080-78eb-11eb-87aa-88dc1254a838.png)

Your help would be much appreciated <3

**Comment by iosifnicolae2** [comment]
@waterdrop01 did you managed to find a workaround? I'm having the same problem.

**Comment by ondrejspilkaABB** [comment]
Hello @tirumaraiselvan, maybe you can provide estimation, when this is going to be fixed?
Use-case is for example batch upserts, so here your product loses some competitiveness...

**Comment by iosifnicolae2** [comment]
I've managed to find a workaround by using a custom function. Also, by using a custom function I was able to do batch processing and to optimise the insert queries quite a lot (10X improvement in ingestion throughout).

**Comment by ondrejspilkaABB** [comment]
@iosifnicolae2 sounds promising, could you provide some details?
Or, is the function accepting array of custom type?

**Comment by iosifnicolae2** [comment]
@ondrejspilkaABB  sure.
So, I've defined a SQL function named `ingest_packets` and registered it in Hasura to be able to use it later via the GraphQL API.
```sql
CREATE OR REPLACE FUNCTION unit.ingest_packets(_data jsonb)
 RETURNS SETOF unit.packet
 LANGUAGE plpgsql
AS $function$
DECLARE
    unit_id         bigint;
    r               jsonb;
    p               jsonb;
    _packet_content jsonb;
BEGIN
    for p in (select jsonb_array_elements((_data ->> 'packets')::jsonb) element)
        loop
            _packet_content := ((p ->> 'packet_content')::jsonb);
            unit_id := unit.insert_or_get_unit((p ->> 'unit_imei')::text);
            
            RETURN NEXT unit.insert_or_get_packet(unit_id, (p ->> 'tcp_session_id')::uuid,
                                                  (_packet_content ->> 'packet')::jsonb);
                                                  
            for r in (select jsonb_array_elements((_packet_content ->> 'records')::jsonb) element)
                loop
                    PERFORM unit.insert_or_update_record(
                            unit_id,
                            (((_packet_content ->> 'packet')::jsonb) ->> 'mongo_id')::text,
                            (((_packet_content ->> 'packet')::jsonb) ->> 'timestamp')::timestamp with time zone,
                            r
                        );
                end loop;
        end loop;
END
$function$
```

and then create `unit.insert_or_get_packet`:
```sql
CREATE OR REPLACE FUNCTION unit.insert_or_get_packet(_unit_id bigint, _session_id uuid, packet jsonb)
 RETURNS SETOF unit.packet
 LANGUAGE plpgsql
AS $function$
BEGIN
    RETURN QUERY INSERT INTO unit.packet (
                                          mongo_id,
                                          session_id,
                                          "timestamp",
                                          data_hex
        )
        VALUES ((packet ->> 'mongo_id')::text,
                _session_id,
                (packet ->> 'timestamp')::timestamp with time zone,
                (packet ->> 'data_hex')::text,) ON CONFLICT (mongo_id, timestamp) DO
            UPDATE
            SET
                mongo_id = EXCLUDED.mongo_id,
                timestamp = EXCLUDED.timestamp,
                session_id = EXCLUDED.session_id,
                data_hex = EXCLUDED.data_hex,
        RETURNING *;
END
$function$
````
and `insert_or_update_record`:
```sql
CREATE OR REPLACE FUNCTION unit.insert_or_update_record(_unit_id bigint, _packet_mongo_id text, _packet_timestamp timestamp with time zone, _record jsonb)
 RETURNS void
 LANGUAGE plpgsql
AS $function$
BEGIN
    INSERT INTO unit.record(unit_id,
                            packet_mongo_id,
                            packet_timestamp,
                            "timestamp",
                            priority,
                            gps_longitude,
                            gps_latitude,
                            gps_altitude,
                            gps_angle,
                            gps_satellites,
                            gps_speed
                            )
    VALUES (_unit_id,
            _packet_mongo_id,
            _packet_timestamp,
            (_record ->> 'timestamp')::timestamp with time zone,
            (_record ->> 'priority')::integer,
            (_record ->> 'gps_longitude')::double precision,
            (_record ->> 'gps_latitude')::double precision,
            (_record ->> 'gps_altitude')::double precision,
            (_record ->> 'gps_angle')::double precision,
            (_record ->> 'gps_satellites')::integer,
            (_record ->> 'gps_speed')::double precision,
            )
    ON CONFLICT (packet_mongo_id, packet_timestamp, timestamp) DO UPDATE
        SET
            unit_id = excluded.unit_id,
            priority = excluded.priority,
            gps_longitude = excluded.gps_longitude,
            gps_latitude = excluded.gps_latitude,
            gps_altitude = excluded.gps_altitude,
            gps_angle = excluded.gps_angle,
            gps_satellites = excluded.gps_satellites,
            gps_speed = excluded.gps_speed;
END
$function$
```
Using the below improvements, I was able to increase the [Bringes](https://bringes.io/) ingestion rate from <50 packets/second to ~5000 packets/second.

Hope this is useful for you. If you have any questions please let me know.

**Comment by sulliwane** [comment]
Thanks @iosifnicolae2 that is super useful.

Would love to see this officially supported by Hasura...

@0x777 any progress on the issue?

**Comment by waterdrop01** [comment]
Wow @0x777, very excited to see that you guys are starting working on this issue :+1: 

I'd be more than happy to help testing any beta/alpha version...

**Comment by waterdrop01** [comment]
Hello @0x777, any chance you share with us the progress and problems you may encounter related to this issue? This would be useful for people following this issue to get a sense of the amount of work remaining to tackle it! Thanks!

**Comment by waterdrop01** [comment]
Hello @0x777, would you mind sharing ongoing discussions regarding this issue? Thanks

**Comment by pe224** [comment]
Until the upstream TimescaleDB issue is fixed, would it be possible to allow `ON CONFLICT DO NOTHING` in graphQL without explicitly specifying any constraint? As the [Postgres docs](https://www.postgresql.org/docs/14/sql-insert.html) state:
> For ON CONFLICT DO NOTHING, it is optional to specify a conflict_target; when omitted, conflicts with all usable constraints (and unique indexes) are handled.

This does not cover the `ON CONFLICT DO UPDATE` case, but might be a helpful first step.

**Comment by moodgorning** [comment]
We have this issue too, and I would like to add a +1 to pe224's suggestion, just having an on conflict do nothing option would already be immensely helpful. We are currently averaging a request error rate of over 30% because of this. Due to the nature of our use case, we can't avoid sending duplicate information to the DB and are using the DB as a deduplication filter. With that, a unique constraint violation should just be ignored, but produces an error with the current functionality instead.

**Comment by ryanbooz** [comment]
I just wanted to leave a message and confirm that this has been fixed in TimescaleDB 2.8 and I just verified through the Hasura API Explorer that the repro steps from @waterdrop01 above now work and successfully UPSERT rows in the hypertable.

The work for this fix was tracked on #1094. 🙌

**Comment by dameleney** [comment]
Thanks for the update @ryanbooz closing this out per your comment that you can no longer reproduce.
https://github.com/hasura/graphql-engine/issues/1711#issuecomment-1267138707
